sdks/java/core/src/main/java/org/apache/beam/sdk/io/datastore/V1Beta3.java

```diff
@@ -166,104 +155,215 @@
+
+    /**
+     * Note that {@code projectId} is only {@code @Nullable} as a matter of build order, but if
+     * it is {@code null} at instantiation time, an error will be thrown.
+     */
+    public Write(@Nullable String projectId) {
+      this.projectId = projectId;
     }
 
     /**
-     * Returns a new {@link Source} that reads the results of the specified query.
-     *
-     * <p>Does not modify this object.
-     *
-     * <p><b>Note:</b> Normally, a Cloud Dataflow job will read from Cloud Datastore in parallel
-     * across many workers. However, when the {@link Query} is configured with a limit using
-     * {@link com.google.datastore.v1beta3.Query.Builder#setLimit(Int32Value)}, then all
-     * returned results will be read by a single Dataflow worker in order to ensure correct data.
+     * Returns a new {@link Write} that writes to the Cloud Datastore for the specified project.
      */
-    public Source withQuery(Query query) {
-      checkNotNull(query, "query");
-      checkArgument(!query.hasLimit() || query.getLimit().getValue() > 0,
-          "Invalid query limit %s: must be positive", query.getLimit().getValue());
-      return new Source(projectId, query, namespace);
+    public Write withProjectId(String projectId) {
+      checkNotNull(projectId, "projectId");
+      return new Write(projectId);
+    }
+
+    @Override
+    public PDone apply(PCollection<Entity> input) {
+      return input.apply(
+          org.apache.beam.sdk.io.Write.to(new DatastoreSink(projectId)));
+    }
+
+    @Override
+    public void validate(PCollection<Entity> input) {
+      checkNotNull(projectId, "projectId");
     }
 
-    public Source withNamespace(@Nullable String namespace) {
-      return new Source(projectId, query, namespace);
+    @Nullable
+    public String getProjectId() {
+      return projectId;
     }
 
     @Override
+    public String toString() {
+      return MoreObjects.toStringHelper(getClass())
+          .add("projectId", projectId)
+          .toString();
+    }
+
+    @Override
+    public void populateDisplayData(DisplayData.Builder builder) {
+      super.populateDisplayData(builder);
+      builder
+          .addIfNotNull(DisplayData.item("projectId", projectId)
+              .withLabel("Output Project"));
+    }
+  }
+
+  /**
+   * A {@link org.apache.beam.sdk.io.Source} that reads data from Datastore.
+   */
+  static class DatastoreSource extends BoundedSource<Entity> {
+
+    @Override
     public Coder<Entity> getDefaultOutputCoder() {
       return ProtoCoder.of(Entity.class);
     }
 
     @Override
     public boolean producesSortedKeys(PipelineOptions options) {
-      // TODO: Perhaps this can be implemented by inspecting the query.
       return false;
     }
 
     @Override
-    public List<Source> splitIntoBundles(long desiredBundleSizeBytes, PipelineOptions options)
-        throws Exception {
+    public List<DatastoreSource> splitIntoBundles(long desiredBundleSizeBytes,
```

Comment:
The class itself package-private. The javadoc will not show this class (and hence the method). The only way this is accessed is through the BoundedSource methods. So I guess it should be fine, unless I am missing something? Are you saying it would be better to make it return List<? extends BoundedSource<Entity>> to make it clear that this isn't a public class?